### ISSUE TITLE
Questions regarding the dependencies

### ISSUE THREAD

**Issue by RobWalt** [Stoeoef/spade#89]
Hey everyone here!

First of all: I love spade! It's the best solution to CDT I could find, it works really reliably and offers a nice API 🤩

Ok ... enough fanboying for now! 😤

I'm currently working on [a PR that integrates `spade` into `geo`](https://github.com/georust/geo/pull/1083). Since the maintainers of `geo` are trying to keep the dependencies clean, I tried to aim for the same with the PR. I noticed the following:

- `spade` uses `robust` 0.2.0, hence I wanted to ask if there's a specific reason that this is the case? Could it be feasible to bump the version?
- `spade` uses the `optional` crate. I'm not really sure what it provides beyond the `std::Option`. The last update to the crate was also 5 years ago. They even mention in [their docs.rs](https://docs.rs/optional/latest/optional/#booleans) that some of the crates use cases was already implemented in the compiler in the meantime. So is there any specific reason (besides the burden of a massive refactoring 😭) this crate is used in spade? Otherwise I could try to get my hands dirty by replacing all of the `optional` stuff with `std::Option`

**Comment by RobWalt** [comment]
I ran the benchmarks with an updated version of `robust` and this is how it turned out

<details>
<summary>

How it turned out

</summary>

```
    Updating crates.io index
   Compiling robust v1.1.0
   Compiling spade v2.2.0 (/home/aviac/repos/clones/spade)
    Finished bench [optimized] target(s) in 4.45s
     Running benches/benchmarks.rs (target/release/deps/benchmarks-a2cdad69c23d9e7a)
Gnuplot not found, using plotters backend
locate benchmark (uniform)/locate (hierarchy<02>), f64
                        time:   [1.0390 µs 1.0433 µs 1.0487 µs]
                        change: [-3.1754% -2.5581% -1.9238%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 50 measurements (18.00%)
  4 (8.00%) high mild
  5 (10.00%) high severe
locate benchmark (uniform)/locate (hierarchy<03>), f64
                        time:   [757.44 ns 758.75 ns 760.61 ns]
                        change: [-2.4910% -2.1193% -1.7443%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 50 measurements (12.00%)
  2 (4.00%) high mild
  4 (8.00%) high severe
locate benchmark (uniform)/locate (hierarchy<04>), f64
                        time:   [701.63 ns 702.58 ns 703.89 ns]
                        change: [-1.4086% -1.2566% -1.0724%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 50 measurements (12.00%)
  1 (2.00%) low mild
  1 (2.00%) high mild
  4 (8.00%) high severe
locate benchmark (uniform)/locate (hierarchy<05>), f64
                        time:   [652.09 ns 654.77 ns 657.42 ns]
                        change: [-0.5443% -0.3202% -0.0958%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 50 measurements (16.00%)
  3 (6.00%) low severe
  2 (4.00%) low mild
  1 (2.00%) high mild
  2 (4.00%) high severe
locate benchmark (uniform)/locate (hierarchy<08>), f64
                        time:   [583.83 ns 584.77 ns 585.94 ns]
                        change: [-4.0957% -3.6904% -3.3875%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 50 measurements (14.00%)
  1 (2.00%) high mild
  6 (12.00%) high severe
locate benchmark (uniform)/locate (hierarchy<10>), f64
                        time:   [567.85 ns 568.34 ns 569.05 ns]
                        change: [-2.7677% -2.6368% -2.5117%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 50 measurements (16.00%)
  4 (8.00%) high mild
  4 (8.00%) high severe
locate benchmark (uniform)/locate (hierarchy<14>), f64
                        time:   [573.91 ns 574.07 ns 574.24 ns]
                        change: [-4.0256% -3.9618% -3.9001%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 50 measurements (4.00%)
  1 (2.00%) high mild
  1 (2.00%) high severe
locate benchmark (uniform)/locate (hierarchy<16>), f64
                        time:   [564.87 ns 565.76 ns 566.93 ns]
                        change: [-6.2166% -6.0429% -5.8788%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 50 measurements (16.00%)
  1 (2.00%) low severe
  1 (2.00%) high mild
  6 (12.00%) high severe
locate benchmark (uniform)/locate (hierarchy<18>), f64
                        time:   [556.56 ns 557.19 ns 558.07 ns]
                        change: [-2.6074% -2.3682% -2.1088%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 50 measurements (12.00%)
  6 (12.00%) high severe
locate benchmark (uniform)/locate (hierarchy<20>), f64
                        time:   [578.70 ns 578.80 ns 578.91 ns]
                        change: [+1.4016% +1.5863% +1.7366%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 50 measurements (10.00%)
  1 (2.00%) low mild
  3 (6.00%) high mild
  1 (2.00%) high severe
locate benchmark (uniform)/locate (hierarchy<32>), f64
                        time:   [558.77 ns 559.61 ns 560.69 ns]
                        change: [-4.0357% -3.7589% -3.4979%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 50 measurements (14.00%)
  1 (2.00%) high mild
  6 (12.00%) high severe
locate benchmark (uniform)/locate (hierarchy<40>), f64
                        time:   [569.45 ns 569.77 ns 570.10 ns]
                        change: [-2.6463% -2.4852% -2.3428%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 50 measurements (8.00%)
  2 (4.00%) high mild
  2 (4.00%) high severe
locate benchmark (uniform)/locate (hierarchy<50>), f64
                        time:   [604.94 ns 605.18 ns 605.50 ns]
                        change: [+0.1108% +0.1981% +0.2991%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high mild
locate benchmark (uniform)/locate (hierarchy<64>), f64
                        time:   [631.32 ns 635.26 ns 639.10 ns]
                        change: [+0.9490% +1.2602% +1.5141%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 50 measurements (8.00%)
  3 (6.00%) low severe
  1 (2.00%) low mild
locate benchmark (uniform)/locate (hierarchy<70>), f64
                        time:   [621.30 ns 621.74 ns 622.33 ns]
                        change: [-3.2505% -3.1369% -2.9512%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 50 measurements (10.00%)
  5 (10.00%) high severe
locate benchmark (uniform)/locate (hierarchy<80>), f64
                        time:   [636.78 ns 637.18 ns 637.58 ns]
                        change: [-2.6870% -2.3972% -2.1598%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  1 (2.00%) low mild
  1 (2.00%) high mild
  1 (2.00%) high severe
locate benchmark (uniform)/locate (hierarchy<90>), f64
                        time:   [643.30 ns 643.77 ns 644.40 ns]
                        change: [-3.5335% -3.2997% -3.0569%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 50 measurements (18.00%)
  2 (4.00%) low mild
  1 (2.00%) high mild
  6 (12.00%) high severe

bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/1000
                        time:   [140.40 µs 140.52 µs 140.69 µs]
                        thrpt:  [7.1078 Melem/s 7.1166 Melem/s 7.1225 Melem/s]
                 change:
                        time:   [-0.3408% -0.1692% +0.0018%] (p = 0.06 > 0.05)
                        thrpt:  [-0.0018% +0.1695% +0.3419%]
                        No change in performance detected.
bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/2000
                        time:   [290.04 µs 290.31 µs 290.65 µs]
                        thrpt:  [6.8812 Melem/s 6.8891 Melem/s 6.8957 Melem/s]
                 change:
                        time:   [+0.8705% +1.0114% +1.1847%] (p = 0.00 < 0.05)
                        thrpt:  [-1.1709% -1.0013% -0.8630%]
                        Change within noise threshold.
Found 3 outliers among 50 measurements (6.00%)
  2 (4.00%) high mild
  1 (2.00%) high severe
bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/3000
                        time:   [523.87 µs 525.69 µs 527.26 µs]
                        thrpt:  [5.6898 Melem/s 5.7068 Melem/s 5.7267 Melem/s]
                 change:
                        time:   [-8.3809% -8.1528% -7.9396%] (p = 0.00 < 0.05)
                        thrpt:  [+8.6244% +8.8765% +9.1476%]
                        Performance has improved.
bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/6000
                        time:   [1.2544 ms 1.2547 ms 1.2550 ms]
                        thrpt:  [4.7809 Melem/s 4.7821 Melem/s 4.7831 Melem/s]
                 change:
                        time:   [-8.4969% -8.2528% -8.0240%] (p = 0.00 < 0.05)
                        thrpt:  [+8.7240% +8.9952% +9.2859%]
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  1 (2.00%) high mild
  2 (4.00%) high severe
bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/10000
                        time:   [2.1162 ms 2.1168 ms 2.1174 ms]
                        thrpt:  [4.7228 Melem/s 4.7241 Melem/s 4.7254 Melem/s]
                 change:
                        time:   [-6.8115% -6.7238% -6.6332%] (p = 0.00 < 0.05)
                        thrpt:  [+7.1045% +7.2085% +7.3094%]
                        Performance has improved.
Found 4 outliers among 50 measurements (8.00%)
  2 (4.00%) high mild
  2 (4.00%) high severe
bulk load benchmark (small)/bulk loading on last used vertex heuristic, locally clustered/15000
                        time:   [3.2779 ms 3.2830 ms 3.2888 ms]
                        thrpt:  [4.5610 Melem/s 4.5690 Melem/s 4.5762 Melem/s]
                 change:
                        time:   [-4.8584% -4.6451% -4.4581%] (p = 0.00 < 0.05)
                        thrpt:  [+4.6662% +4.8714% +5.1065%]
                        Performance has improved.
Found 2 outliers among 50 measurements (4.00%)
  2 (4.00%) high mild
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/1000
                        time:   [135.42 µs 135.51 µs 135.60 µs]
                        thrpt:  [7.3746 Melem/s 7.3796 Melem/s 7.3844 Melem/s]
                 change:
                        time:   [-0.4208% -0.2989% -0.1867%] (p = 0.00 < 0.05)
                        thrpt:  [+0.1870% +0.2998% +0.4226%]
                        Change within noise threshold.
Found 3 outliers among 50 measurements (6.00%)
  3 (6.00%) high mild
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/2000
                        time:   [280.74 µs 281.03 µs 281.34 µs]
                        thrpt:  [7.1089 Melem/s 7.1166 Melem/s 7.1241 Melem/s]
                 change:
                        time:   [+0.1279% +0.2238% +0.3134%] (p = 0.00 < 0.05)
                        thrpt:  [-0.3124% -0.2233% -0.1278%]
                        Change within noise threshold.
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/3000
                        time:   [480.56 µs 483.38 µs 486.27 µs]
                        thrpt:  [6.1694 Melem/s 6.2062 Melem/s 6.2428 Melem/s]
                 change:
                        time:   [-10.114% -9.7576% -9.3987%] (p = 0.00 < 0.05)
                        thrpt:  [+10.374% +10.813% +11.251%]
                        Performance has improved.
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/6000
                        time:   [1.1552 ms 1.1580 ms 1.1614 ms]
                        thrpt:  [5.1660 Melem/s 5.1814 Melem/s 5.1941 Melem/s]
                 change:
                        time:   [-7.6734% -7.5325% -7.3645%] (p = 0.00 < 0.05)
                        thrpt:  [+7.9500% +8.1462% +8.3111%]
                        Performance has improved.
Found 4 outliers among 50 measurements (8.00%)
  2 (4.00%) high mild
  2 (4.00%) high severe
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/10000
                        time:   [2.0145 ms 2.0224 ms 2.0289 ms]
                        thrpt:  [4.9288 Melem/s 4.9447 Melem/s 4.9641 Melem/s]
                 change:
                        time:   [-7.3774% -7.1767% -6.9730%] (p = 0.00 < 0.05)
                        thrpt:  [+7.4957% +7.7315% +7.9650%]
                        Performance has improved.
bulk load benchmark (small)/bulk loading on last used vertex heuristic, uniformly distributed/15000
                        time:   [3.1493 ms 3.1538 ms 3.1581 ms]
                        thrpt:  [4.7497 Melem/s 4.7562 Melem/s 4.7629 Melem/s]
                 change:
                        time:   [-7.8462% -7.6751% -7.5238%] (p = 0.00 < 0.05)
                        thrpt:  [+8.1359% +8.3132% +8.5143%]
                        Performance has improved.
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/1000
                        time:   [147.83 µs 147.95 µs 148.06 µs]
                        thrpt:  [6.7541 Melem/s 6.7588 Melem/s 6.7647 Melem/s]
                 change:
                        time:   [-0.5465% -0.2175% +0.3244%] (p = 0.46 > 0.05)
                        thrpt:  [-0.3233% +0.2180% +0.5495%]
                        No change in performance detected.
Found 2 outliers among 50 measurements (4.00%)
  1 (2.00%) high mild
  1 (2.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/2000
                        time:   [310.80 µs 311.02 µs 311.28 µs]
                        thrpt:  [6.4250 Melem/s 6.4304 Melem/s 6.4351 Melem/s]
                 change:
                        time:   [+1.5532% +1.6119% +1.6679%] (p = 0.00 < 0.05)
                        thrpt:  [-1.6405% -1.5864% -1.5294%]
                        Performance has regressed.
Found 5 outliers among 50 measurements (10.00%)
  3 (6.00%) high mild
  2 (4.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/3000
                        time:   [560.72 µs 561.80 µs 563.22 µs]
                        thrpt:  [5.3265 Melem/s 5.3400 Melem/s 5.3503 Melem/s]
                 change:
                        time:   [-6.8008% -6.5305% -6.2001%] (p = 0.00 < 0.05)
                        thrpt:  [+6.6099% +6.9868% +7.2970%]
                        Performance has improved.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high mild
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/6000
                        time:   [1.2978 ms 1.2999 ms 1.3027 ms]
                        thrpt:  [4.6060 Melem/s 4.6157 Melem/s 4.6231 Melem/s]
                 change:
                        time:   [-9.5702% -9.3869% -9.2058%] (p = 0.00 < 0.05)
                        thrpt:  [+10.139% +10.359% +10.583%]
                        Performance has improved.
Found 2 outliers among 50 measurements (4.00%)
  2 (4.00%) high mild
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/10000
                        time:   [2.3252 ms 2.3278 ms 2.3305 ms]
                        thrpt:  [4.2909 Melem/s 4.2959 Melem/s 4.3007 Melem/s]
                 change:
                        time:   [-7.9695% -7.3499% -6.9047%] (p = 0.00 < 0.05)
                        thrpt:  [+7.4168% +7.9330% +8.6596%]
                        Performance has improved.
Found 2 outliers among 50 measurements (4.00%)
  1 (2.00%) high mild
  1 (2.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, uniformly distributed/15000
                        time:   [3.6288 ms 3.6328 ms 3.6379 ms]
                        thrpt:  [4.1233 Melem/s 4.1291 Melem/s 4.1336 Melem/s]
                 change:
                        time:   [-6.0168% -5.9383% -5.8578%] (p = 0.00 < 0.05)
                        thrpt:  [+6.2223% +6.3132% +6.4019%]
                        Performance has improved.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high mild
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/1000
                        time:   [152.89 µs 153.21 µs 153.56 µs]
                        thrpt:  [6.5121 Melem/s 6.5269 Melem/s 6.5405 Melem/s]
                 change:
                        time:   [-1.3543% -1.1393% -0.9463%] (p = 0.00 < 0.05)
                        thrpt:  [+0.9554% +1.1524% +1.3729%]
                        Change within noise threshold.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high mild
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/2000
                        time:   [314.27 µs 315.04 µs 315.88 µs]
                        thrpt:  [6.3315 Melem/s 6.3484 Melem/s 6.3639 Melem/s]
                 change:
                        time:   [+1.9198% +2.0699% +2.2484%] (p = 0.00 < 0.05)
                        thrpt:  [-2.1990% -2.0280% -1.8836%]
                        Performance has regressed.
Found 5 outliers among 50 measurements (10.00%)
  1 (2.00%) high mild
  4 (8.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/3000
                        time:   [582.79 µs 583.55 µs 584.35 µs]
                        thrpt:  [5.1339 Melem/s 5.1410 Melem/s 5.1476 Melem/s]
                 change:
                        time:   [-8.7711% -8.6400% -8.4983%] (p = 0.00 < 0.05)
                        thrpt:  [+9.2876% +9.4571% +9.6144%]
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  2 (4.00%) low mild
  1 (2.00%) high mild
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/6000
                        time:   [1.4077 ms 1.4087 ms 1.4099 ms]
                        thrpt:  [4.2556 Melem/s 4.2594 Melem/s 4.2624 Melem/s]
                 change:
                        time:   [-7.3952% -7.2530% -7.0514%] (p = 0.00 < 0.05)
                        thrpt:  [+7.5864% +7.8202% +7.9858%]
                        Performance has improved.
Found 9 outliers among 50 measurements (18.00%)
  7 (14.00%) high mild
  2 (4.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/10000
                        time:   [2.3845 ms 2.3903 ms 2.3967 ms]
                        thrpt:  [4.1723 Melem/s 4.1835 Melem/s 4.1937 Melem/s]
                 change:
                        time:   [-6.3118% -6.1920% -6.0859%] (p = 0.00 < 0.05)
                        thrpt:  [+6.4803% +6.6008% +6.7370%]
                        Performance has improved.
Found 8 outliers among 50 measurements (16.00%)
  3 (6.00%) low severe
  2 (4.00%) low mild
  2 (4.00%) high mild
  1 (2.00%) high severe
bulk load benchmark (small)/bulk loading on hierarchy lookup, locally clustered/15000
                        time:   [3.6799 ms 3.6870 ms 3.6949 ms]
                        thrpt:  [4.0596 Melem/s 4.0684 Melem/s 4.0762 Melem/s]
                 change:
                        time:   [-4.4747% -4.3173% -4.1599%] (p = 0.00 < 0.05)
                        thrpt:  [+4.3404% +4.5122% +4.6843%]
                        Performance has improved.
Found 9 outliers among 50 measurements (18.00%)
  8 (16.00%) low mild
  1 (2.00%) high mild

bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/100000
                        time:   [22.332 ms 22.371 ms 22.430 ms]
                        thrpt:  [4.4583 Melem/s 4.4702 Melem/s 4.4779 Melem/s]
                 change:
                        time:   [-7.2350% -7.0682% -6.8491%] (p = 0.00 < 0.05)
                        thrpt:  [+7.3527% +7.6058% +7.7993%]
                        Performance has improved.
Found 16 outliers among 50 measurements (32.00%)
  7 (14.00%) low severe
  5 (10.00%) low mild
  4 (8.00%) high severe
bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/200000
                        time:   [45.102 ms 45.113 ms 45.126 ms]
                        thrpt:  [4.4321 Melem/s 4.4333 Melem/s 4.4344 Melem/s]
                 change:
                        time:   [-6.3477% -5.9415% -5.6368%] (p = 0.00 < 0.05)
                        thrpt:  [+5.9735% +6.3168% +6.7779%]
                        Performance has improved.
Benchmarking bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/400000: Warming up for 1.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 5.1s, or reduce sample count to 40.
bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/400000
                        time:   [101.62 ms 101.69 ms 101.76 ms]
                        thrpt:  [3.9309 Melem/s 3.9337 Melem/s 3.9364 Melem/s]
                 change:
                        time:   [-5.7098% -5.3291% -5.0380%] (p = 0.00 < 0.05)
                        thrpt:  [+5.3052% +5.6291% +6.0556%]
                        Performance has improved.
Found 2 outliers among 50 measurements (4.00%)
  2 (4.00%) high mild
Benchmarking bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/800000: Warming up for 1.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 11.2s, or reduce sample count to 20.
bulk load benchmark (big)/bulk loading on last used vertex heuristic, uniformly distributed/800000
                        time:   [223.39 ms 223.56 ms 223.76 ms]
                        thrpt:  [3.5753 Melem/s 3.5785 Melem/s 3.5812 Melem/s]
                 change:
                        time:   [-5.4819% -5.1896% -4.9259%] (p = 0.00 < 0.05)
                        thrpt:  [+5.1811% +5.4737% +5.7998%]
                        Performance has improved.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high severe

bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/1000
                        time:   [146.07 µs 146.11 µs 146.16 µs]
                        thrpt:  [6.8420 Melem/s 6.8443 Melem/s 6.8459 Melem/s]
                 change:
                        time:   [-0.8233% -0.5601% -0.3116%] (p = 0.00 < 0.05)
                        thrpt:  [+0.3125% +0.5633% +0.8301%]
                        Change within noise threshold.
bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/2000
                        time:   [306.05 µs 306.21 µs 306.42 µs]
                        thrpt:  [6.5269 Melem/s 6.5315 Melem/s 6.5350 Melem/s]
                 change:
                        time:   [+0.6029% +0.6871% +0.7888%] (p = 0.00 < 0.05)
                        thrpt:  [-0.7826% -0.6824% -0.5993%]
                        Change within noise threshold.
Found 5 outliers among 50 measurements (10.00%)
  3 (6.00%) high mild
  2 (4.00%) high severe
bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/3000
                        time:   [553.19 µs 553.74 µs 554.27 µs]
                        thrpt:  [5.4125 Melem/s 5.4177 Melem/s 5.4231 Melem/s]
                 change:
                        time:   [-10.112% -9.9960% -9.8782%] (p = 0.00 < 0.05)
                        thrpt:  [+10.961% +11.106% +11.250%]
                        Performance has improved.
Found 1 outliers among 50 measurements (2.00%)
  1 (2.00%) high mild
bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/6000
                        time:   [1.3075 ms 1.3076 ms 1.3078 ms]
                        thrpt:  [4.5877 Melem/s 4.5884 Melem/s 4.5890 Melem/s]
                 change:
                        time:   [-8.7490% -8.6914% -8.6474%] (p = 0.00 < 0.05)
                        thrpt:  [+9.4660% +9.5187% +9.5878%]
                        Performance has improved.
Found 4 outliers among 50 measurements (8.00%)
  2 (4.00%) high mild
  2 (4.00%) high severe
bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/10000
                        time:   [2.3186 ms 2.3190 ms 2.3194 ms]
                        thrpt:  [4.3115 Melem/s 4.3122 Melem/s 4.3129 Melem/s]
                 change:
                        time:   [-6.8707% -6.8013% -6.7305%] (p = 0.00 < 0.05)
                        thrpt:  [+7.2162% +7.2976% +7.3776%]
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  2 (4.00%) high mild
  1 (2.00%) high severe
bulk vs incremental loading/bulk loading on hierarchy lookup, uniformly distributed/15000
                        time:   [3.6127 ms 3.6140 ms 3.6159 ms]
                        thrpt:  [4.1484 Melem/s 4.1505 Melem/s 4.1520 Melem/s]
                 change:
                        time:   [-6.6256% -6.5140% -6.3954%] (p = 0.00 < 0.05)
                        thrpt:  [+6.8324% +6.9679% +7.0957%]
                        Performance has improved.
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/1000
                        time:   [223.97 µs 224.04 µs 224.10 µs]
                        thrpt:  [4.4622 Melem/s 4.4635 Melem/s 4.4648 Melem/s]
                 change:
                        time:   [-2.1549% -2.0734% -1.9887%] (p = 0.00 < 0.05)
                        thrpt:  [+2.0290% +2.1173% +2.2024%]
                        Performance has improved.
Found 5 outliers among 50 measurements (10.00%)
  4 (8.00%) high mild
  1 (2.00%) high severe
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/2000
                        time:   [625.36 µs 625.90 µs 626.48 µs]
                        thrpt:  [3.1924 Melem/s 3.1954 Melem/s 3.1982 Melem/s]
                 change:
                        time:   [-7.0696% -6.8312% -6.5976%] (p = 0.00 < 0.05)
                        thrpt:  [+7.0636% +7.3321% +7.6075%]
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  2 (4.00%) high mild
  1 (2.00%) high severe
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/3000
                        time:   [1.1713 ms 1.1725 ms 1.1738 ms]
                        thrpt:  [2.5558 Melem/s 2.5586 Melem/s 2.5613 Melem/s]
                 change:
                        time:   [-7.5192% -7.3817% -7.2498%] (p = 0.00 < 0.05)
                        thrpt:  [+7.8165% +7.9700% +8.1305%]
                        Performance has improved.
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/6000
                        time:   [2.6563 ms 2.6571 ms 2.6580 ms]
                        thrpt:  [2.2573 Melem/s 2.2581 Melem/s 2.2588 Melem/s]
                 change:
                        time:   [-5.6075% -5.5424% -5.4710%] (p = 0.00 < 0.05)
                        thrpt:  [+5.7877% +5.8677% +5.9406%]
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  3 (6.00%) high severe
Benchmarking bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/10000: Warming up for 1.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 6.2s, enable flat sampling, or reduce sample count to 30.
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/10000
                        time:   [4.7693 ms 4.7723 ms 4.7766 ms]
                        thrpt:  [2.0935 Melem/s 2.0954 Melem/s 2.0967 Melem/s]
                 change:
                        time:   [-4.0226% -3.7930% -3.5925%] (p = 0.00 < 0.05)
                        thrpt:  [+3.7264% +3.9425% +4.1911%]
                        Performance has improved.
Benchmarking bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/15000: Warming up for 1.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 9.9s, enable flat sampling, or reduce sample count to 20.
bulk vs incremental loading/incremental loading on hierarchy lookup, uniformly distributed/15000
                        time:   [7.6391 ms 7.6400 ms 7.6411 ms]
                        thrpt:  [1.9631 Melem/s 1.9633 Melem/s 1.9636 Melem/s]
                 change:
                        time:   [-5.1273% -4.9606% -4.8107%] (p = 0.00 < 0.05)
                        thrpt:  [+5.0538% +5.2195% +5.4044%]
                        Performance has improved.
Found 8 outliers among 50 measurements (16.00%)
  2 (4.00%) low mild
  2 (4.00%) high mild
  4 (8.00%) high severe
```

</details>

I would argue that it is a solid ~5-10% performance improvement for free. A few benchmarks also regressed (like 4/50 or so) and I would just blame that on bad rng.

**Comment by Stoeoef** [comment]
> Hey everyone here!
> 
> First of all: I love spade! It's the best solution to CDT I could find, it works really reliably and offers a nice API 🤩
> 
> Ok ... enough fanboying for now! 😤
> 
> I'm currently working on [a PR that integrates `spade` into `geo`](https://github.com/georust/geo/pull/1083). Since the maintainers of `geo` are trying to keep the dependencies clean, I tried to aim for the same with the PR. I noticed the following:
> 
>     * `spade` uses `robust` 0.2.0, hence I wanted to ask if there's a specific reason that this is the case? Could it be feasible to bump the version?
> 
>     * `spade` uses the `optional` crate. I'm not really sure what it provides beyond the `std::Option`. The last update to the crate was also 5 years ago. They even mention in [their docs.rs](https://docs.rs/optional/latest/optional/#booleans) that some of the crates use cases was already implemented in the compiler in the meantime. So is there any specific reason (besides the burden of a massive refactoring 😭) this crate is used in spade? Otherwise I could try to get my hands dirty by replacing all of the `optional` stuff with `std::Option`

Thanks for the kind words :)
Originally I've introduced `optional` to reduce the size of `Option<FixedEdgeHandle>` from 5 bytes (handle = 4 bytes, Option adds 1) to 4 bytes. This should, theoretically, make a difference when storing vertices and edges  - see structs `FaceEntry` and `VertexEntry`. But since the measurements don't show a significant difference and the actual size reduction won't be large I'm happy with removing the dependency!

**Comment by Stoeoef** [comment]
Fixed with version 2.2.1, available on https://crates.io/crates/spade/2.2.1
Thanks again for your contributions!
